internal/client/client.go

```diff
@@ -122,124 +161,124 @@ func (c *CSAPI) SendEventSynced(t *testing.T, roomID string, e b.Event) string {
-			event.Get("state_key").Str == c.UserID
-	})
-}
-
-// SyncUntil blocks and continually calls /sync until
-// - the response contains a particular `key`, and
-// - its corresponding value is an array
-// - some element in that array makes the `check` function return true.
-// If the `check` function fails the test, the failing event will be automatically logged.
-// Will time out after CSAPI.SyncUntilTimeout.
+// Incremental /sync example: (test controls since token)
+//    since := alice.MustSyncUntil(t, client.SyncReq{TimeoutMillis: "0"}) // get a since token
+//    bob.InviteRoom(t, roomID, alice.UserID)
+//    since = alice.MustSyncUntil(t, client.SyncReq{Since: since}, client.SyncInvitedTo(alice.UserID, roomID))
+//    alice.JoinRoom(t, roomID, nil)
+//    alice.MustSyncUntil(t, client.SyncReq{Since: since}, client.SyncJoinedTo(alice.UserID, roomID))
 //
-// Returns the `next_batch` token from the last /sync response. This can be passed as
-// `since` to sync from this point forward only.
-func (c *CSAPI) SyncUntil(t *testing.T, since, filter, key string, check func(gjson.Result) bool) string {
+// Checking multiple parts of /sync:
+//    alice.MustSyncUntil(
+//        t, client.SyncReq{},
+//        client.SyncJoinedTo(alice.UserID, roomID),
+//        client.SyncJoinedTo(alice.UserID, roomID2),
+//        client.SyncJoinedTo(alice.UserID, roomID3),
+//    )
+//
+// Check functions are unordered and independent. Once a check function returns true it is removed
+// from the list of checks and won't be called again.
+//
+// In the unlikely event that you want all the checkers to pass *explicitly* in a single /sync
+// response (e.g to assert some form of atomic update which updates multiple parts of the /sync
+// response at once) then make your own checker function which does this.
+//
+// In the unlikely event that you need ordering on your checks, call MustSyncUntil multiple times
+// with a single checker, and reuse the returned since token, as in the "Incremental sync" example.
+//
+// Will time out after CSAPI.SyncUntilTimeout. Returns the latest since token used.
+func (c *CSAPI) MustSyncUntil(t *testing.T, syncReq SyncReq, checks ...SyncCheckOpt) string {
 	t.Helper()
 	start := time.Now()
-	checkCounter := 0
-	// Print failing events in a defer() so we handle t.Fatalf in the same way as t.Errorf
-	var wasFailed = t.Failed()
-	var lastEvent *gjson.Result
-	timedOut := false
-	defer func() {
-		if !wasFailed && t.Failed() {
-			raw := ""
-			if lastEvent != nil {
-				raw = lastEvent.Raw
-			}
-			if !timedOut {
-				t.Logf("SyncUntil: failing event %s", raw)
-			}
+	numResponsesReturned := 0
+	checkers := make([]struct {
+		check SyncCheckOpt
+		errs  []string
+	}, len(checks))
+	for i := range checks {
+		c := checkers[i]
+		c.check = checks[i]
+		checkers[i] = c
+	}
+	printErrors := func() string {
+		err := "Checkers:\n"
+		for _, c := range checkers {
+			err += strings.Join(c.errs, "\n")
+			err += ", "
```

Comment:
```suggestion
			err += ", \n"
```

It looks like we need to add a new-line to the end of each batch of `checker` errors. Currently, if there are multiple failing `checkers`, then there will be some overlap between them


```go
alice.MustSyncUntil(
	t, client.SyncReq{},
	client.SyncJoinedTo(alice.UserID, roomID),
	client.SyncJoinedTo(alice.UserID, "!some-random-room"),
	client.SyncJoinedTo(alice.UserID, "!no-exist-room"),
)
```
			

```diff
    msc2716_test.go:96: @alice:hs1 MustSyncUntil: timed out after 5.082929509s. Seen 6 /sync responses. Checkers:
        [t=41.077935ms] Response #1: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
        [t=1.047872777s] Response #2: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
        [t=2.056341503s] Response #3: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
        [t=3.067336659s] Response #4: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
        [t=4.07435841s] Response #5: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
-        [t=5.082922736s] Response #6: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist, [t=41.090211ms] Response #1: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
        [t=1.047877957s] Response #2: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
        [t=2.056346991s] Response #3: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
        [t=3.067342371s] Response #4: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
        [t=4.07436217s] Response #5: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
        [t=5.082928211s] Response #6: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist,
```

---

To make it more clear, we could also add grouping/separation:

(mocked up)

```
    msc2716_test.go:96: @alice:hs1 MustSyncUntil: timed out after 5.082929509s. Seen 6 /sync responses. Checkers:
        Failed checker #1:
            [t=41.077935ms] Response #1: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
            [t=1.047872777s] Response #2: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
            [t=2.056341503s] Response #3: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
            [t=3.067336659s] Response #4: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
            [t=4.07435841s] Response #5: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
            [t=5.082922736s] Response #6: SyncJoinedTo(@alice:hs1,!some-random-room): SyncTimelineHas(!some-random-room): Key rooms.join.!some-random-room.timeline.events does not exist
        Failed checker #2:
            [t=41.090211ms] Response #1: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
            [t=1.047877957s] Response #2: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
            [t=2.056346991s] Response #3: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
            [t=3.067342371s] Response #4: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
            [t=4.07436217s] Response #5: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
            [t=5.082928211s] Response #6: SyncJoinedTo(@alice:hs1,!no-exist-room): SyncTimelineHas(!no-exist-room): Key rooms.join.!no-exist-room.timeline.events does not exist
```